### PR TITLE
Adding well control equations

### PR DESCRIPTION
**Pull request by GitPaean**
This is part of the efforts to have proper VFP support. At the stage, it only recovers to the extent of the VFP capacity of the `flow_legacy`, while it not suitable for more general THP/VFP treatment yet. 

Please do not merge before requesting. 

`Adding well control equations` is the major part, basically, it reduces the complexity of the functions `getQs` and `updateWellstate`, and also, it provides functionality to monitor the convergence of the well control equations.

### CODE REVIEW

**Comment by GitPaean** [comment]
#1477 provides a better fix to recover the VFP/THP capacity of `flow_legacy`. This PR will adapt to that fixes.

**Comment by totto82** [comment]
Is that other benefits to adding the control equations separately than improving the VFP/THP results? I guess the performance penalty is small since it just adds one more equation pr well, but have you checked?

**Comment by GitPaean** [comment]
The performance is impacted very slightly, it will be more related to iteration behavior. Some cases is faster (SPE9_solvent, if I remember correctly), while some cases is slightly slower.  It was meant to recover the old `VFP/THP` results, but it turned out that one was due how to initialize the perforation density when use it to calculate the BHP from THP, as we both discovered.

The major benefit will be to reduce the complexity of the well code. For example, for functions `getQs`, we have always need to check the current control types, to decide how to get well rates, which results in very complicated code. With this PR, we just get them directly from the primary variables. `getBhp` is also much simpler, and also `updateWellState`. Basically, after we initialized the primary variables, we get most information from primary variables directly. Basically, the code is better formulated and also the well equation system. 

And also, it introduces possibility to check the convergence of the control equations. If you plot the well curves from this PR and https://github.com/OPM/opm-simulators/pull/1477, results both look good, while well curves from this PR is smoother, since we enforce the convergence of the control equation here. We can do that with separate checking, while it was not as well formulated. 

And also, what I hoped, with the introduction of an extra equation to describe the relation between the BHP (primary variables), THP(targets) and well rates (from primary variables), all the variables will be more synchronized/coupled, which results in better convergence hopefully or we can enforce stricter convergence criterion, while we do not see that from SPE1 based THP/VFP tests.

**Comment by GitPaean** [comment]
I will update this PR after  #1477 goes in.

File: opm/autodiff/StandardWell.hpp [hunk 1]
```diff
@@ -57,15 +57,30 @@ namespace Opm
...
+        // correspondingly, we have four well equations for blackoil model, the first three are mass
+        // converstation equations, and the last one is the well control equation.
+        // primary variables related to other components, will be before the Bhp and after F_g.
+        // well control equation is always the last well equation.
+        // TODO: in the current implementation, we use the well rate as the first primary variables for injectors,
+        // instead of G_t.
         static const bool gasoil = numEq == 2 && (Indices::compositionSwitchIdx >= 0);
-        static const int XvarWell = 0;
+        static const int GTotal = 0;
         static const int WFrac = gasoil? -1000: 1;
```
Comment (totto82) [inline opm/autodiff/StandardWell.hpp]: Do we have a more descriptive name for this (these) variable?
Could this work?
```
GTotal -> weightedTotalRate
WFrac -> waterFraction etc. 
```

File: opm/autodiff/StandardWell_impl.hpp [hunk 2]
```diff
@@ -35,7 +35,7 @@ namespace Opm
     , perf_pressure_diffs_(number_of_perforations_)
     , primary_variables_(numWellEq, 0.0)
     , primary_variables_evaluation_(numWellEq) // the number of the primary variables
-    , F0_(numWellEq)
+    , F0_(num_components_)
     {
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: Why did you change this? It should be number of equations.

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: Nope. It only applies to mass balance equations, not well control equation. It should be `numWellEq - 1`  or `num_components_`.

File: opm/autodiff/StandardWell.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/autodiff/StandardWell.hpp]: It will work. In my opinion, they are frequently used indices, and kind of good to keep them short.  It is okay to give comments about their definition when defining them. 

Anyway, I do not have strong feeling about this. If you insist, I will do a search-replace through the file, or in another PR. 

Just let me know how you think.

File: opm/autodiff/StandardWell_impl.hpp [hunk 3]
```diff
@@ -122,176 +119,59 @@ namespace Opm
...
+                comp_frac = comp_frac_[pu.phase_pos[ Gas ]] * wsolvent();
+            } else if (legacyCompIdx == pu.phase_pos[ Gas ]) {
+                comp_frac = comp_frac_[legacyCompIdx];
+                if (has_solvent) {
+                    comp_frac *= (1.0 - wsolvent());
                 }
-                return (target_rate * wellVolumeFractionScaled(comp_idx) / combined_volume_fraction);
+            } else {
+                comp_frac = comp_frac_[legacyCompIdx];
             }
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: Why not just use WellInjector::TypeEnum injectorType from the parser. 
```
If(type == injectorType::Water && comp_idx != Water
 || type == injectorType::Oil && comp_idx != Oil)
||   type == injectorType::Gas && comp_idx != Gas )
   return 0
else 
  return  primary_variables_evaluation_[GTotal]
```

Comment (totto82) [reply opm/autodiff/StandardWell_impl.hpp]: I think this is a good opportunity to start getting rid of comp_frac and distr !

File: opm/autodiff/StandardWell_impl.hpp [hunk 4]
```diff
@@ -731,11 +611,115 @@ namespace Opm
...
+                    if (distr[phase] > 0.0) {
+                        ++number_phases_under_control;
+                    }
+                }
+                assert(number_phases_under_control > 0);
+
+                const double target_rate = well_controls_get_current_target(well_controls_); // reservoir rate target
+                if (well_type_ == INJECTOR) {
+                    assert(number_phases_under_control == 1);
+                    for (int phase = 0; phase < number_of_phases_; ++phase) {
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: Why not use `assert(injectortype =! WellInjector::MULTI)`

File: opm/autodiff/StandardWell_impl.hpp [hunk 5]
```diff
@@ -855,34 +839,71 @@ namespace Opm
...
+
+    template<typename TypeTag>
+    void
+    StandardWell<TypeTag>::
+    processFractions() const
+    {
         assert(FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx));
+        const auto pu = phaseUsage();
+        std::vector<double> F(number_of_phases_, 0.0);
         F[pu.phase_pos[Oil]] = 1.0;
```
Comment (totto82) [inline opm/autodiff/StandardWell_impl.hpp]: Are we still using the legacy phase position for F[...]?

File: opm/autodiff/StandardWell_impl.hpp (hunk 2 above)
Comment (totto82) [reply opm/autodiff/StandardWell_impl.hpp]: Ok. I get it. numWellEq == num_components + num_control_eq. Then I agree on your change.

Comment (totto82) [reply opm/autodiff/StandardWell_impl.hpp]: So now we have for 
: blackoil 
num_components = 3 numWelleq = 4
: polymer
num_components = 3 numWellEq = 4 
: solvent
num_componets = 4 numWellEq = 4

Maybe it is better with an extra variable` numWellConservationEq` 
Then you get: 
```
num_components 
numWellConservationEq (== num_components for all but the polymer case) 
numWellEq (== numWellConservationEq + 1)
```

The current implementation (also before this PR) are too often mixing the concept of number of components, phases and equations.

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: > : solvent
num_componets = 4 numWellEq = 4

So we do not have a mass balance equation for the solvent model?  The solvent part is something difficult for me, since I do not understand the details.


> The current implementation (also before this PR) are too often mixing the concept of number of components, phases and equations.

It has been always been this way.  We should make clear that what is the equation index, what is the primary variable index. (component/phase thing is a little harder and I do not have good suggestions, but we should provide a fix/convention eventually).

**Comment by totto82** [comment]
I think this should be merged, when properly tested, especially on all the different variants we have: 
oilwater, oilgas, solvent, polymer and thermal. I went for the current variable switching system when prototyping the well equations for the frankenstein simulator in order to have the same number of equations in the well and the reservoir. We are no longer restricted to that, so I am fine with adding the control equation explicitly to make the implementation simpler.

File: opm/autodiff/StandardWell.hpp (hunk 1 above)
Comment (totto82) [reply opm/autodiff/StandardWell.hpp]: I see your point. No strong opinions on my side either. But what is the "G" in GTotal?

**Comment by GitPaean** [comment]
Thanks for the review, I will address them later.  I am trying to improving the situation that have a zero phase rate limit. There were some fixes in https://github.com/OPM/opm-simulators/pull/1414, while with this PR, something different needs to be done.

File: opm/autodiff/StandardWell.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/autodiff/StandardWell.hpp]: I do not recall the reason, probably I felt `Q` is not very correct since it is not the total well rate, then I used `G`. Maybe I should use `WQTotal` to say it is `weighted Q total`.

**Comment by GitPaean** [comment]
`zero phase rate` control is taken care of. in #1414, there is another situation taken care of while has not been taken care of here. It is that, a well can not produce a phase, while we specify a target for that phase, basically it is mission impossible. Not sure how to detect it yet.  The last commit improves the running with Norne. Different convergence behavoir can be caused that we have an extra equation (control equation) to check convergence.

**Comment by GitPaean** [comment]
For norne, it is very slightly slower (less than 1% or something).  (please pay attention to the iteration numbers, the timing is always fluctuating a little bit.)
Master branch
```
61240 ================    End of simulation     ===============
61241 
61242 Total time (seconds):         1158.44
61243 Solver time (seconds):        1154
61244  Assembly time (seconds):     667.213 (Failed: 22.9794; 3.44408%)
61245  Linear solve time (seconds): 455.397 (Failed: 16.9395; 3.71972%)
61246  Update time (seconds):       19.1142 (Failed: 0.711832; 3.7241%)
61247  Output write time (seconds): 5.73246
61248 Overall Well Iterations:      977 (Failed: 21; 2.14944%)
61249 Overall Linearizations:       1948 (Failed: 66; 3.38809%)
61250 Overall Newton Iterations:    1609 (Failed: 66; 4.10193%)
61251 Overall Linear Iterations:    24455 (Failed: 894; 3.65569%)
```

with this PR
```
61205 ================    End of simulation     ===============
61206 
61207 Total time (seconds):         1123.71
61208 Solver time (seconds):        1119.44
61209  Assembly time (seconds):     632.457 (Failed: 25.0094; 3.95432%)
61210  Linear solve time (seconds): 456.586 (Failed: 20.3417; 4.45519%)
61211  Update time (seconds):       18.3983 (Failed: 0.792854; 4.30938%)
61212  Output write time (seconds): 5.53953
61213 Overall Well Iterations:      978 (Failed: 22; 2.24949%)
61214 Overall Linearizations:       1954 (Failed: 77; 3.94063%)
61215 Overall Newton Iterations:    1614 (Failed: 77; 4.77076%)
61216 Overall Linear Iterations:    24581 (Failed: 1087; 4.42211%)
```

For model 2, it is also slightly slower (1% or something)
With mater branch
```
171119 
171120 Total time (seconds):         4176.85
171121 Solver time (seconds):        4151.45
171122  Assembly time (seconds):     1840.37 (Failed: 132.316; 7.18964%)
171123  Linear solve time (seconds): 2222.11 (Failed: 114.597; 5.15714%)
171124  Update time (seconds):       63.6785 (Failed: 4.43897; 6.9709%)
171125  Output write time (seconds): 28.2826
171126 Overall Well Iterations:      752 (Failed: 38; 5.05319%)
171127 Overall Linearizations:       2156 (Failed: 153; 7.09647%)
171128 Overall Newton Iterations:    1822 (Failed: 153; 8.39737%)
171129 Overall Linear Iterations:    46989 (Failed: 2248; 4.7841%)
```

with this PR
```
171213 
171214 Total time (seconds):         4356.69
171215 Solver time (seconds):        4331.18
171216  Assembly time (seconds):     1928.19 (Failed: 158.051; 8.19687%)
171217  Linear solve time (seconds): 2319.19 (Failed: 154.63; 6.66745%)
171218  Update time (seconds):       58.8949 (Failed: 5.09634; 8.65328%)
171219  Output write time (seconds): 28.5105
171220 Overall Well Iterations:      779 (Failed: 41; 5.26316%)
171221 Overall Linearizations:       2184 (Failed: 174; 7.96703%)
171222 Overall Newton Iterations:    1845 (Failed: 174; 9.43089%)
171223 Overall Linear Iterations:    47134 (Failed: 2981; 6.32452%)
171224 
```

File: opm/autodiff/StandardWell_impl.hpp (hunk 3 above)
Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: Removing `comp_frac` and `distr` is better to be put in a separate PR, since it has been used many places and we better limit the scope of the current PR. Especially, we had something unsolved as reported in https://github.com/OPM/opm-simulators/pull/1441 , we need to do that in a separate PR for better testing.

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]:
```
If(type == injectorType::Water && comp_idx != Water
 || type == injectorType::Oil && comp_idx != Oil)
||   type == injectorType::Gas && comp_idx != Gas )
   return 0
else 
  return  primary_variables_evaluation_[GTotal]
```

solvent part is part I am always a little nervous about. I am not sure how to incorporate solvent related here.

**Comment by GitPaean** [comment]
I think we should limit the PR to be focused on its purpose. I mean, different tasks should be separated into different PRs, so it will make both the testing and development easier. 

For this PR, removing the usage of `comp_frac` and `distr`, and introducing the `WellInectionProperties` and `WellProducerProperties` will be pretty significant efforts and they should be addressed with three different PRs, and will not be covered with this PR. Especially, there is some difficulties that might prevent us from doing that.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
https://github.com/OPM/opm-simulators/pull/1480 will fix the inversion failure problem. But it is one modification will impact almost all the simulations, we need to address https://github.com/OPM/opm-simulators/pull/1480 in a good way separately. Please help to suggest and comment about https://github.com/OPM/opm-simulators/pull/1480 so we can make it in soon. Then we can come back to this PR.

**Comment by GitPaean** [comment]
As @andlaus  suggested, adding a line of code 
```
 Dune::FMatrixPrecision<K>::set_singular_limit(1.e-20);
``` 
to lower the singular limit of the invert() function.  It is possibly the simplest fix for the problem. And it also means, the specialized `4X4` invert method should be removed, since that failure was from the same reason.

**Comment by GitPaean** [comment]
jenkins buid this please

**Comment by GitPaean** [comment]
> jenkins buid this please

**Comment by GitPaean** [comment]
jenkins buid this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=18 please

**Comment by blattms** [comment]
> As @andlaus  suggested, adding a line of code 
> ```
>  Dune::FMatrixPrecision<K>::set_singular_limit(1.e-20);
> ``` 
> to lower the singular limit of the invert() function.  It is possibly the simplest fix for the problem. And it also means, the specialized `4X4` invert method should be removed, since that failure was from the same reason. 

Are you sure that the specialized versions do not use less floating point operations?

**Comment by GitPaean** [comment]
> Are you sure that the specialized versions do not use less floating point operations?

I am not sure. It was introduced to handle failure in inverting. And since it is one matrix per well, I am not worrying about the performance impact from this part of the simulation. Simpler code and better readability/maintainability will be more important. 

Although I am not totally sure, getting the LU decomposition and reusing it to solve linear system is very likely more expensive than inverting it once and reuse the inverse of the matrix, especially for the function `addWellContribution`, but we basically all agree we should avoid to do the inverse, although how to it needs to be discussed.  `Lapack` looks like the current option to go, since there is some work to do to make the luDecompostion of DenseMatrix publicly available.

Anyway, current status of the PR does not introduce extra inversion failure now.  Whether to use inverse of not will be a general improvements and not related to this PR directly.

**Comment by GitPaean** [comment]
jenkins build this opm-common=401 opm-tests=18 please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=18 please

**Comment by GitPaean** [comment]
With `jenkins build this opm-tests=18 please`, 

```
Initialized empty Git repository in /var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-tests/.git/
From https://github.com/OPM/opm-tests
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> origin/master
From https://github.com/OPM/opm-tests
 * [new ref]         refs/pull/18/merge -> branch_to_build
Already on 'master'
Branch master set up to track remote branch master from origin.
fatal: refusing to merge unrelated histories
Build step 'Execute shell' marked build as failure
Skipped archiving because build is not successful
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
```

**Comment by GitPaean** [comment]
jenkins build this opm-common=401 opm-tests=18 please

**Comment by GitPaean** [comment]
sorry. I should not have tried `opm-common=401`, somehow, I `"saw"` that PR is not merged.

**Comment by akva2** [comment]
401 is the problem, we have to revert it. works now, please quell the introduced warning.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=18 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=18 please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=18 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=21 please

**Comment by GitPaean** [comment]
Not sure why the one #18 generated with `update_data` failed here. Anyway, the manually created #21 works well.

**Comment by GitPaean** [comment]
I think the change of the way of the volume calculation changes the iteration behavior a little.  The following is the testing results with this PR and without this PR for Norne and Model 2.  Again, please pay attention to the iteration numbers instead of timing. 

Norne. 
With this PR
```
61141 
61142 Total time (seconds):         1196.56
61143 Solver time (seconds):        1191
61144  Assembly time (seconds):     648.143 (Failed: 18.8191; 2.90355%)
61145  Linear solve time (seconds): 507.638 (Failed: 16.3118; 3.21327%)
61146  Update time (seconds):       21.2169 (Failed: 0.681559; 3.21234%)
61147  Output write time (seconds): 7.01738
61148 Overall Well Iterations:      969 (Failed: 15; 1.54799%)
61149 Overall Linearizations:       1930 (Failed: 55; 2.84974%)
61150 Overall Newton Iterations:    1592 (Failed: 55; 3.45477%)
61151 Overall Linear Iterations:    23999 (Failed: 759; 3.16263%)
61152 
61153 
61154 
61155 Error summary:
61156 Warnings          281
61157 Info              1087
61158 Errors            0
61159 Bugs              0
61160 Debug             3762
61161 Problems          5
```
Without this PR
```
61227 
61228 Total time (seconds):         1078.16
61229 Solver time (seconds):        1073.77
61230  Assembly time (seconds):     586.444 (Failed: 20.2752; 3.45731%)
61231  Linear solve time (seconds): 457.747 (Failed: 17.4286; 3.80748%)
61232  Update time (seconds):       17.6603 (Failed: 0.659195; 3.73263%)
61233  Output write time (seconds): 5.63145
61234 Overall Well Iterations:      977 (Failed: 21; 2.14944%)
61235 Overall Linearizations:       1940 (Failed: 66; 3.40206%)
61236 Overall Newton Iterations:    1601 (Failed: 66; 4.12242%)
61237 Overall Linear Iterations:    24318 (Failed: 912; 3.75031%)
61238 
61239 
61240 
61241 Error summary:
61242 Warnings          337
61243 Info              1095
61244 Errors            0
61245 Bugs              0
61246 Debug             3778
61247 Problems          6
61248 
```

Model 2
With this PR
```
171131 Total time (seconds):         4370.67
171132 Solver time (seconds):        4354.26
171133  Assembly time (seconds):     1807.67 (Failed: 141.221; 7.81231%)
171134  Linear solve time (seconds): 2465.17 (Failed: 144.654; 5.86792%)
171135  Update time (seconds):       55.8649 (Failed: 4.59544; 8.226%)
171136  Output write time (seconds): 19.2649
171137 Overall Well Iterations:      769 (Failed: 41; 5.3316%)
171138 Overall Linearizations:       2187 (Failed: 166; 7.59031%)
171139 Overall Newton Iterations:    1850 (Failed: 166; 8.97297%)
171140 Overall Linear Iterations:    47581 (Failed: 2583; 5.42864%)
171141 
171142 
171143 
171144 Error summary:
171145 Warnings          242
171146 Info              1024
171147 Errors            0
171148 Bugs              0
171149 Debug             4055
171150 Problems          16
171151 
```
Without this PR
```
171231 Total time (seconds):         4615.8
171232 Solver time (seconds):        4597.66
171233  Assembly time (seconds):     1968.91 (Failed: 185.19; 9.40573%)
171234  Linear solve time (seconds): 2528.36 (Failed: 214.587; 8.4872%)
171235  Update time (seconds):       71.7115 (Failed: 6.45162; 8.99663%)
171236  Output write time (seconds): 21.4461
171237 Overall Well Iterations:      783 (Failed: 49; 6.25798%)
171238 Overall Linearizations:       2214 (Failed: 207; 9.34959%)
171239 Overall Newton Iterations:    1873 (Failed: 207; 11.0518%)
171240 Overall Linear Iterations:    48847 (Failed: 3960; 8.10695%)
171241 
171242 
171243 
171244 Error summary:
171245 Warnings          242
171246 Info              1035
171247 Errors            0
171248 Bugs              0
171249 Debug             4112
171250 Problems          22
```

**Comment by totto82** [comment]
From my point of view this can be merged now, or do you still have some WIP stuff (the PR title). I will merge it later today if nobody objects.

**Comment by GitPaean** [comment]
> From my point of view this can be merged now, or do you still have some WIP stuff (the PR title). I will merge it later today if nobody objects.

Thanks for the comment. This PR is kind of a relatively independent change that impacts almost all the simulations and should be tested and merged by itself. Other development will be proposed as separate PRs. So please go ahead to merge when agreed.

**Comment by totto82** [comment]
Thanks. Merging now as promised.
